kuri_camera/src/tcp_test_client.cpp

```diff
@@ -0,0 +1,137 @@
+#include <ros/ros.h>
+#include <boost/asio.hpp>
+#include <iostream>
+#include <mutex>
+#include <thread>
+#include <boost/thread.hpp>
+#include <condition_variable>
+
+void async_io_service_run(boost::asio::io_service& io_service) {
+  ROS_INFO("async_io_service_run");
+  boost::thread t(boost::bind(&boost::asio::io_service::run, &io_service));
+  t.detach();
+}
+
+void deestablishConnection(boost::asio::ip::tcp::socket& socket, boost::asio::io_service& io_service) {
+  socket.close();
+  io_service.reset();
+  async_io_service_run(io_service);
+}
+
+void establishConnection(boost::asio::ip::tcp::socket& socket, const boost::asio::ip::tcp::endpoint& endpoint, boost::asio::io_service& io_service) {
+
+  std::mutex connectMutex;
+  std::condition_variable connectCvar;
+
+  ros::Rate retryRate(1.0);
+  boost::system::error_code error = boost::asio::error::host_not_found;
+
+  while (ros::ok() && error) {
+    std::unique_lock<std::mutex> connectLock(connectMutex);
+    socket.async_connect(endpoint, [&error, &connectCvar, &connectMutex](const boost::system::error_code& err) {
+      // ROS_ERROR("In connect handler");
+      std::unique_lock<std::mutex> connectLock2(connectMutex);
+      error = err;
+      connectLock2.unlock();
+      connectCvar.notify_all();
+    });
+    // async_io_service_run(io_service);
+    std::cv_status timerResult = connectCvar.wait_for(connectLock, std::chrono::seconds(5));
+    if (timerResult == std::cv_status::timeout) {
+      ROS_ERROR("Connect Timeout");
+      error = boost::asio::error::timed_out;
+    } else {
+      // ROS_ERROR("Connect No timeout");
+    }
+    if (error) {
+      ROS_ERROR("Error connecting, will keep trying. Error: %s", error.message().c_str());
+      deestablishConnection(socket, io_service);
+      retryRate.sleep();
+    }
+    connectLock.unlock();
+  }
+}
+
+int main(int argc, char **arcv) {
+  ros::init(argc, arcv, "tcp_client");
+  ros::Time::init();
+
+  boost::asio::io_service io_service;
+  // boost::asio::io_service::work work(io_service);
+  boost::asio::ip::tcp::socket socket(io_service);
+  boost::asio::ip::tcp::resolver resolver(io_service);
+
+  std::string tcpSocketHostname;
+  int tcpSocketPort;
+  ros::param::param<std::string>("tcpSocketHostname", tcpSocketHostname, "cococutkuri.personalrobotics.cs.washington.edu");
```

Comment:
Doesn't matter much for testing code, but would leave this without a default so that people aren't surprised (they should get a "no arg provided" not a "cannot resolve mysterious hostname")

Reply:
Done